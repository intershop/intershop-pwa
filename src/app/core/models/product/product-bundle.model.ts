import { SkuQuantityType } from './product.helper';
import { Product } from './product.model';

export interface ProductBundle extends Product {
  bundledProducts: SkuQuantityType[];
}
