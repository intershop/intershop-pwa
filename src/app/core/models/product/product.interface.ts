import { AttributeGroup } from 'ish-core/models/attribute-group/attribute-group.model';
import { Attribute } from 'ish-core/models/attribute/attribute.model';
import { CategoryData } from 'ish-core/models/category/category.interface';
import { Image } from 'ish-core/models/image/image.model';
import { Link } from 'ish-core/models/link/link.model';
import { PriceData } from 'ish-core/models/price/price.interface';
import { VariationAttribute } from 'ish-core/models/product-variation/variation-attribute.model';
import { SeoAttributesData } from 'ish-core/models/seo-attributes/seo-attributes.interface';
import { Warranty } from 'ish-core/models/warranty/warranty.model';

export interface ProductData {
  sku: string;
  productName: string;
  shortDescription: string;
  longDescription: string;

  availability: boolean;
  inStock: boolean;
  availableStock?: number;

  averageRating: string;
  roundedAverageRating: string;

  images: Image[];

  availableWarranties?: Warranty[];
  availableGiftWraps?: unknown[];
  availableGiftMessages?: unknown[];
  bundles: unknown[];
  retailSet: boolean;

  // If warranty {
  price?: PriceData;
  currency?: unknown;
  // }

  productMasterSKU?: string;
  minOrderQuantity: number;
  maxOrderQuantity?: number;
  stepOrderQuantity?: number;
  packingUnit: string;

  // If Variation Master and Retail Set {
  minListPrice?: PriceData;
  maxListPrice?: PriceData;
  minSalePrice?: PriceData;
  maxSalePrice?: PriceData;
  // }
  variationAttributeValues?: VariationAttribute[];
  variableVariationAttributes?: VariationAttribute[];
  partOfRetailSet: boolean;
  // If  Retail Set {
  summedUpListPrice?: PriceData;
  summedUpSalePrice?: PriceData;
  // }

  attachments?: unknown;
  variations?: unknown;
  crosssells?: unknown;
  productMaster: boolean;
  productBundle: boolean;
  listPrice: PriceData;
  salePrice: PriceData;
  manufacturer: string;
  mastered: boolean;
  readyForShipmentMin: number;
  readyForShipmentMax: number;
  attributes?: Attribute[];
  productTypes: string[];
  attributeGroups?: { [id: string]: AttributeGroup };
  defaultCategory?: CategoryData;

  promotions?: Link[];

  seoAttributes: SeoAttributesData;
}

export interface ProductDataStub {
  attributeGroup?: { name: string; attributes: Attribute[] };
  attributes: Attribute[];
  description: string;
  title: string;
}

export interface ProductVariationLink extends Link {
  variableVariationAttributeValues: VariationAttribute[];
}
