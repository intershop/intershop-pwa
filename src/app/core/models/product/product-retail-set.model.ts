import { Price } from 'ish-core/models/price/price.model';

import { Product } from './product.model';

export interface ProductRetailSet extends Product {
  minListPrice: Price;
  minSalePrice: Price;
  summedUpListPrice: Price;
  summedUpSalePrice: Price;
  partSKUs: string[];
}
