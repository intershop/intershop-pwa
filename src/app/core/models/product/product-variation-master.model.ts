import { Price } from 'ish-core/models/price/price.model';
import { VariationAttribute } from 'ish-core/models/product-variation/variation-attribute.model';

import { Product } from './product.model';

export interface VariationProductMaster extends Product {
  variationAttributeValues?: VariationAttribute[];
  variationSKUs?: string[];
  defaultVariationSKU?: string;
  minListPrice?: Price;
  minSalePrice?: Price;
  maxListPrice?: Price;
  maxSalePrice?: Price;
}
