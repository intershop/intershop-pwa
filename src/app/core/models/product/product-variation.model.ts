import { VariationAttribute } from 'ish-core/models/product-variation/variation-attribute.model';

import { Product } from './product.model';

export interface VariationProduct extends Product {
  variableVariationAttributes?: VariationAttribute[];
  productMasterSKU?: string;
}
