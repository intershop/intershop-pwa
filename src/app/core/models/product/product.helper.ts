import { intersection } from 'lodash-es';

import { Attribute } from 'ish-core/models/attribute/attribute.model';
import { Image } from 'ish-core/models/image/image.model';
import { PriceHelper } from 'ish-core/models/price/price.model';
import {
  ProductView,
  VariationProductMasterView,
  VariationProductView,
} from 'ish-core/models/product-view/product-view.model';

import { ProductBundle } from './product-bundle.model';
import { ProductRetailSet } from './product-retail-set.model';
import { VariationProductMaster } from './product-variation-master.model';
import { VariationProduct } from './product-variation.model';
import { Product } from './product.model';

export interface SkuQuantityType {
  sku: string;
  quantity: number;
}

export enum ProductCompletenessLevel {
  Detail = 3,
  List = 2,
}

export type AllProductTypes = Product | VariationProduct | VariationProductMaster | ProductBundle | ProductRetailSet;

export type AnyProductViewType = ProductView | VariationProductView | VariationProductMasterView;

// not-dead-code
export type ProductPrices = Partial<
  Pick<ProductRetailSet, 'minListPrice' | 'minSalePrice' | 'summedUpListPrice' | 'summedUpSalePrice'>
> &
  Partial<Pick<VariationProductMaster, 'minListPrice' | 'minSalePrice' | 'maxListPrice' | 'maxSalePrice'>> &
  Partial<Pick<Product, 'salePrice' | 'listPrice'>>;

export class ProductHelper {
  /**
   * Get primary product image based on image type
   * @param product   The Product for which to get the primary image
   * @param imageType The wanted ImageType
   * @returns         The primary product image of the given ImageType
   */
  static getPrimaryImage(product: Product, imageType: string): Image {
    if (!(product && product.images)) {
      return;
    }
    return product.images.find(image => image.typeID === imageType && image.primaryImage);
  }

  /**
   * Get product image based on image type and image view
   * @param product   The Product for which to get the image
   * @param imageType The wanted ImageType
   * @param imageView The wanted ImageView
   * @returns         The matching product image
   */
  static getImageByImageTypeAndImageView(product: Product, imageType: string, imageView: string): Image {
    if (!(product && product.images)) {
      return;
    }
    return product.images.find(image => image.typeID === imageType && image.viewID === imageView);
  }

  /**
   * Get all product ImageView ids matching image type
   * @param product   The Product for which to get the image types
   * @param imageType The wanted ImageType
   * @returns         Array of available ImageView ids
   */
  static getImageViewIDs(product: Product, imageType: string): string[] {
    if (!(product && product.images)) {
      return [];
    }
    return product.images
      .filter(image => image.typeID === imageType)
      .sort((a, b) => (a.primaryImage ? -1 : b.primaryImage ? 1 : 0))
      .map(image => image.viewID);
  }

  /**
   * check if a given product has a sufficient completeness level
   */
  static isSufficientlyLoaded(product: Product, completenessLevel: ProductCompletenessLevel): boolean {
    return !!product && product.completenessLevel >= completenessLevel;
  }

  static isFailedLoading(product: Product): boolean {
    return !!product && !!product.failed;
  }

  static isReadyForDisplay(product: Product, completenessLevel: ProductCompletenessLevel) {
    return ProductHelper.isSufficientlyLoaded(product, completenessLevel) || ProductHelper.isFailedLoading(product);
  }

  /**
   * Check if product is a retail set
   */
  static isRetailSet(product: Partial<AllProductTypes>): product is ProductRetailSet {
    return product && product.type === 'RetailSet';
  }

  /**
   * Check if product is a master product
   */
  static isMasterProduct(
    product: Partial<AllProductTypes>
  ): product is VariationProductMaster & VariationProductMasterView {
    return product && product.type === 'VariationProductMaster';
  }

  /**
   * Check if product is a variation product
   */
  static isVariationProduct(product: Partial<AllProductTypes>): product is VariationProduct & VariationProductView {
    return product && product.type === 'VariationProduct';
  }

  /**
   * Check if product is a product bundle
   */
  static isProductBundle(product: Partial<AllProductTypes>): product is ProductBundle {
    return product && product.type === 'Bundle';
  }

  /**
   * Get product attributes by attribute group id.
   * @param product           The Product for which to get the attributes
   * @param attributeGroupId  The attribute group id of the attributes to get
   * @returns                 The product attributes of the attribute group (if any)
   */
  static getAttributesOfGroup(product: Product, attributeGroupId: string): Attribute[] {
    if (
      product &&
      product.attributeGroups &&
      product.attributeGroups[attributeGroupId] &&
      product.attributeGroups[attributeGroupId].attributes &&
      product.attributeGroups[attributeGroupId].attributes.length > 0
    ) {
      return product.attributeGroups[attributeGroupId].attributes;
    }
    return;
  }

  // not-dead-code
  static calculatePriceRange(products: Product[]): ProductPrices {
    if (!products || !products.length) {
      return {};
    } else if (products.length === 1) {
      return products[0];
    } else {
      return {
        minListPrice: products.map(p => p.listPrice).reduce(PriceHelper.min),
        minSalePrice: products.map(p => p.salePrice).reduce(PriceHelper.min),
        summedUpListPrice: products.map(p => p.listPrice).reduce(PriceHelper.sum),
        summedUpSalePrice: products.map(p => p.salePrice).reduce(PriceHelper.sum),
      };
    }
  }

  /**
   * Determines the set of common attribute names for the compare products.
   * @param products List of products to be compared
   * @returns        A set of the common attribute names
   */
  static getCommonAttributeNames(products: Product[]): string[] {
    if (!products || !products.length) {
      return [];
    }
    const result = products.filter(x => !!x).map(product => product.attributes.map(x => x.name));
    return intersection(...result);
  }

  /**
   * Get a product with only specific attributes. All attributes that are common between the compare products are filtered out.
   * @param product         The product that should be stripped of its common attributes
   * @param visibleProducts List of products to be compared
   * @returns               A Product with specific attributes only compared to the common attributes
   */
  static getProductWithoutCommonAttributes(product: Product, visibleProducts: Product[]): Product {
    if (!product || !product.sku || !visibleProducts || !visibleProducts.length) {
      return;
    }
    const common = ProductHelper.getCommonAttributeNames(visibleProducts);
    const attributes = product.attributes.filter(att => !common.includes(att.name));
    return { ...product, attributes };
  }
}
