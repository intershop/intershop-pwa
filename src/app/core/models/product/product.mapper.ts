import { Injectable } from '@angular/core';
import { flatten } from 'lodash-es';

import { AttributeGroup } from 'ish-core/models/attribute-group/attribute-group.model';
import { AttributeHelper } from 'ish-core/models/attribute/attribute.helper';
import { CategoryData } from 'ish-core/models/category/category.interface';
import { CategoryMapper } from 'ish-core/models/category/category.mapper';
import { ImageMapper } from 'ish-core/models/image/image.mapper';
import { Link } from 'ish-core/models/link/link.model';
import { PriceMapper } from 'ish-core/models/price/price.mapper';
import { SeoAttributesMapper } from 'ish-core/models/seo-attributes/seo-attributes.mapper';

import { SkuQuantityType } from './product.helper';
import { ProductData, ProductDataStub, ProductVariationLink } from './product.interface';
import { AllProductTypes, Product, VariationProduct, VariationProductMaster } from './product.model';

/**
 * check if attribute is available and return value, otherwise undefined
 */
function retrieveStubAttributeValue<T>(data: ProductDataStub, attributeName: string) {
  return data ? AttributeHelper.getAttributeValueByAttributeName<T>(data.attributes, attributeName) : undefined;
}

/**
 * maps promotion links data to a string array of promotion ids
 */
function mapPromotionIds(data: Link[]): string[] {
  return data ? data.map(promotionLink => promotionLink.itemId) : [];
}

/**
 * maps an attribute group and its attribute data to the same format as it is used in a single product call
 */
function mapAttributeGroups(data: ProductDataStub): { [id: string]: AttributeGroup } {
  return { [data.attributeGroup.name]: { attributes: data.attributeGroup.attributes } };
}

/**
 * Product Mapper maps data of HTTP requests to model instances and vice versa.
 */
@Injectable({ providedIn: 'root' })
export class ProductMapper {
  constructor(private imageMapper: ImageMapper, private categoryMapper: CategoryMapper) {}

  static parseSKUfromURI(uri: string): string {
    const match = /products[^\/]*\/([^\?]*)/.exec(uri);
    if (match) {
      return match[1];
    } else {
      console.warn(`could not find sku in uri '${uri}'`);
      return;
    }
  }

  static findDefaultVariation(variationLinks: Link[]): string {
    const defaultVariation = variationLinks.find(variation =>
      AttributeHelper.getAttributeValueByAttributeName<boolean>(variation.attributes, 'defaultVariation')
    );

    return defaultVariation ? ProductMapper.parseSKUfromURI(defaultVariation.uri) : undefined;
  }

  static constructMasterStub(sku: string, variations: Partial<VariationProduct>[]): Partial<VariationProductMaster> {
    return (
      variations?.length && {
        type: 'VariationProductMaster',
        sku,
        completenessLevel: 0,
        variationAttributeValues: flatten(variations.map(v => v.variableVariationAttributes)).filter(
          (val, idx, arr) =>
            arr.findIndex(el => el.variationAttributeId === val.variationAttributeId && el.value === val.value) === idx
        ),
      }
    );
  }

  fromLink(link: Link): Partial<Product> {
    return {
      sku: ProductMapper.parseSKUfromURI(link.uri),
      name: link.title,
      shortDescription: link.description,
      type: 'Product',
      completenessLevel: 1,
    };
  }

  fromVariationLink(link: ProductVariationLink, productMasterSKU: string): Partial<VariationProduct> {
    return {
      ...this.fromLink(link),
      variableVariationAttributes: link.variableVariationAttributeValues,
      productMasterSKU,
      type: 'VariationProduct',
      failed: false,
    };
  }

  fromRetailSetLink(link: Link): Partial<Product> {
    return {
      sku: ProductMapper.parseSKUfromURI(link.uri),
      name: link.title,
      shortDescription: link.description,
      completenessLevel: 1,
    };
  }

  private calculateAvailable(availability: boolean, inStock: boolean) {
    return !!availability && (inStock !== undefined ? inStock : true);
  }

  /**
   * construct a {@link Product} stub from data returned by link list responses with additional data
   */
  fromStubData(data: ProductDataStub): Partial<AllProductTypes> {
    const sku = retrieveStubAttributeValue<string>(data, 'sku');
    if (!sku) {
      throw new Error('cannot construct product stub without SKU');
    }
    const productCategory = retrieveStubAttributeValue<CategoryData>(data, 'defaultCategory');

    const promotionLinks = retrieveStubAttributeValue<{ elements: Link[] }>(data, 'promotions')?.elements ?? [];

    const mastered = retrieveStubAttributeValue<boolean>(data, 'mastered');
    const productMaster = retrieveStubAttributeValue<boolean>(data, 'productMaster');
    const productMasterSKU = retrieveStubAttributeValue<string>(data, 'productMasterSKU');
    const retailSet = retrieveStubAttributeValue<boolean>(data, 'retailSet');

    const product: Partial<Product> = {
      shortDescription: data.description,
      name: data.title,
      sku,
      listPrice: PriceMapper.fromData(retrieveStubAttributeValue(data, 'listPrice')),
      salePrice: PriceMapper.fromData(retrieveStubAttributeValue(data, 'salePrice')),
      images: this.imageMapper.fromImages([
        {
          effectiveUrl: retrieveStubAttributeValue(data, 'image'),
          name: 'front M',
          primaryImage: true,
          type: 'Image',
          typeID: 'M',
          viewID: 'front',
          imageActualHeight: undefined,
          imageActualWidth: undefined,
        },
        {
          effectiveUrl: retrieveStubAttributeValue(data, 'image'),
          name: 'front S',
          primaryImage: true,
          type: 'Image',
          typeID: 'S',
          viewID: 'front',
          imageActualHeight: undefined,
          imageActualWidth: undefined,
        },
      ]),
      manufacturer: retrieveStubAttributeValue(data, 'manufacturer'),
      available: this.calculateAvailable(
        retrieveStubAttributeValue(data, 'availability'),
        retrieveStubAttributeValue(data, 'inStock')
      ),
      longDescription: undefined,
      minOrderQuantity: retrieveStubAttributeValue<{ value: number }>(data, 'minOrderQuantity')?.value || 1,
      maxOrderQuantity: retrieveStubAttributeValue<{ value: number }>(data, 'maxOrderQuantity')?.value || 100,
      stepOrderQuantity: retrieveStubAttributeValue<{ value: number }>(data, 'stepOrderQuantity')?.value || 1,
      packingUnit: retrieveStubAttributeValue(data, 'packingUnit'),
      attributeGroups: data.attributeGroup && mapAttributeGroups(data),
      readyForShipmentMin: undefined,
      readyForShipmentMax: undefined,
      roundedAverageRating: +retrieveStubAttributeValue<string>(data, 'roundedAverageRating') || 0,
      type: 'Product',
      defaultCategoryId: productCategory ? this.categoryMapper.fromDataSingle(productCategory).uniqueId : undefined,
      promotionIds: mapPromotionIds(promotionLinks),
      completenessLevel: 2,
      failed: false,
    };

    if (productMaster) {
      return {
        ...product,
        type: 'VariationProductMaster',
      };
    } else if (mastered) {
      return {
        ...product,
        productMasterSKU,
        type: 'VariationProduct',
      };
    } else if (retailSet) {
      return {
        ...product,
        type: 'RetailSet',
      };
    } else {
      return product;
    }
  }

  /**
   * map API Response to fully qualified {@link Product}s
   */
  fromData(data: ProductData): AllProductTypes {
    const product: Product = {
      type: 'Product',
      name: data.productName,
      shortDescription: data.shortDescription,
      longDescription: data.longDescription,
      available: this.calculateAvailable(data.availability, data.inStock),
      minOrderQuantity: data.minOrderQuantity || 1,
      maxOrderQuantity: data.maxOrderQuantity || 100,
      stepOrderQuantity: data.stepOrderQuantity || 1,
      packingUnit: data.packingUnit,
      availableStock: data.availableStock,
      attributes:
        (data.attributeGroups &&
          data.attributeGroups.PRODUCT_DETAIL_ATTRIBUTES &&
          data.attributeGroups.PRODUCT_DETAIL_ATTRIBUTES.attributes) ||
        data.attributes,
      attributeGroups: data.attributeGroups,
      images: this.imageMapper.fromImages(data.images),
      listPrice: PriceMapper.fromData(data.listPrice),
      salePrice: PriceMapper.fromData(data.salePrice),
      manufacturer: data.manufacturer,
      readyForShipmentMin: data.readyForShipmentMin,
      readyForShipmentMax: data.readyForShipmentMax,
      roundedAverageRating: +data.roundedAverageRating || 0,
      sku: data.sku,
      defaultCategoryId: data.defaultCategory
        ? this.categoryMapper.fromDataSingle(data.defaultCategory).uniqueId
        : undefined,
      promotionIds: mapPromotionIds(data.promotions),
      completenessLevel: 3,
      failed: false,
      seoAttributes: SeoAttributesMapper.fromData(data.seoAttributes),
    };

    if (data.productMaster) {
      return {
        ...product,
        minListPrice: PriceMapper.fromData(data.minListPrice),
        minSalePrice: PriceMapper.fromData(data.minSalePrice),
        maxListPrice: PriceMapper.fromData(data.maxListPrice),
        maxSalePrice: PriceMapper.fromData(data.maxSalePrice),
        variationAttributeValues: data.variationAttributeValues,
        type: 'VariationProductMaster',
      };
    } else if (data.mastered) {
      return {
        ...product,
        productMasterSKU: data.productMasterSKU,
        variableVariationAttributes: data.variableVariationAttributes,
        type: 'VariationProduct',
      };
    } else if ((data.productTypes && data.productTypes.includes('BUNDLE')) || data.productBundle) {
      return {
        ...product,
        type: 'Bundle',
      };
    } else if (data.retailSet) {
      return {
        ...product,
        type: 'RetailSet',
        minListPrice: PriceMapper.fromData(data.minListPrice),
        minSalePrice: PriceMapper.fromData(data.minSalePrice),
        summedUpListPrice: PriceMapper.fromData(data.summedUpListPrice),
        summedUpSalePrice: PriceMapper.fromData(data.summedUpSalePrice),
      };
    } else {
      return product;
    }
  }

  fromProductBundleData(links: Link[]): SkuQuantityType[] {
    return links.map(link => ({
      sku: ProductMapper.parseSKUfromURI(link.uri),
      quantity: AttributeHelper.getAttributeValueByAttributeName<{ value: number }>(link.attributes, 'quantity').value,
    }));
  }
}
