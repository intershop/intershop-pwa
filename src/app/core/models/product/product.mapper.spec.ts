import { TestBed } from '@angular/core/testing';
import { provideMockStore } from '@ngrx/store/testing';
import { anything, spy, verify } from 'ts-mockito';

import { Attribute } from 'ish-core/models/attribute/attribute.model';
import { ImageMapper } from 'ish-core/models/image/image.mapper';
import { Link } from 'ish-core/models/link/link.model';
import { getICMBaseURL } from 'ish-core/store/core/configuration';

import { ProductData, ProductDataStub } from './product.interface';
import { ProductMapper } from './product.mapper';
import { Product, ProductHelper, VariationProductMaster } from './product.model';

describe('Product Mapper', () => {
  let productMapper: ProductMapper;
  let imageMapper: ImageMapper;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [provideMockStore({ selectors: [{ selector: getICMBaseURL, value: 'http://www.example.org' }] })],
    });
    productMapper = TestBed.inject(ProductMapper);
    imageMapper = spy(TestBed.inject(ImageMapper));
  });

  describe('fromData', () => {
    it(`should return Product when getting a ProductData`, () => {
      const product: Product = productMapper.fromData({ sku: '1' } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('Product');
      verify(imageMapper.fromImages(anything())).once();
    });

    it(`should return VariationProduct when getting a ProductData with mastered = true`, () => {
      const product: Product = productMapper.fromData({
        sku: '1',
        mastered: true,
        productMasterSKU: '2',
      } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('VariationProduct');
      expect(ProductHelper.isMasterProduct(product)).toBeFalsy();
      verify(imageMapper.fromImages(anything())).once();
    });

    it(`should return VariationProductMaster when getting a ProductData with productMaster = true`, () => {
      const product: Product = productMapper.fromData({ sku: '1', productMaster: true } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('VariationProductMaster');
      expect(ProductHelper.isMasterProduct(product)).toBeTruthy();
    });

    it(`should return VariationProductMaster with variationAttributes when getting a ProductData with productMaster = true`, () => {
      const product: Product = productMapper.fromData({
        sku: '1',
        productMaster: true,
        variationAttributeValues: [],
      } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('VariationProductMaster');
      expect(ProductHelper.isMasterProduct(product)).toBeTruthy();
      expect((product as VariationProductMaster).variationAttributeValues).toBeEmpty();
    });

    it(`should return Product with variationAttributes when getting a ProductData with productMaster = false`, () => {
      const product: Product = productMapper.fromData({
        sku: '1',
        productMaster: false,
        variableVariationAttributes: [],
      } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('Product');
      expect(ProductHelper.isMasterProduct(product)).toBeFalsy();
      expect((product as VariationProductMaster).variationAttributeValues).toBeFalsy();
    });

    it('should return ProductBundle when getting a ProductData with productBundle = true', () => {
      const product: Product = productMapper.fromData({
        sku: '1',
        productBundle: true,
      } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('Bundle');
      expect(ProductHelper.isProductBundle(product)).toBeTrue();
    });

    it('should return ProductBundle when getting a ProductData with producttype contains "BUNDLE"', () => {
      const product: Product = productMapper.fromData({
        sku: '1',
        productTypes: ['BUNDLE'],
      } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('Bundle');
      expect(ProductHelper.isProductBundle(product)).toBeTrue();
    });

    it('should return ProductRetailSet when getting a ProductData with retailSet = true', () => {
      const product: Product = productMapper.fromData({
        sku: '1',
        retailSet: true,
      } as ProductData);
      expect(product).toBeTruthy();
      expect(product.type).toEqual('RetailSet');
      expect(ProductHelper.isRetailSet(product)).toBeTrue();
    });

    it('should return attributes or detailed product attributes when PRODUCT_DETAIL_ATTRIBUTES enabled', () => {
      const p1: Product = productMapper.fromData({
        sku: '1',
        attributes: [{ name: 'Graphics', type: 'String', value: 'NVIDIA Quadro K2200' }],
      } as ProductData);
      const p2: Product = productMapper.fromData(({
        sku: '1',
        attributes: [{ name: 'Graphics', type: 'String', value: 'NVIDIA Quadro K2200' }],
        attributeGroups: {
          PRODUCT_DETAIL_ATTRIBUTES: {
            attributes: [{ name: 'Grafikkarte', type: 'String', value: 'NVIDIA Quadro K2200' }],
          },
        },
      } as unknown) as ProductData);
      expect(p1).toBeTruthy();
      expect(p1.attributes).toEqual([{ name: 'Graphics', type: 'String', value: 'NVIDIA Quadro K2200' }]);
      expect(p2).toBeTruthy();
      expect(p2.attributes).toEqual([{ name: 'Grafikkarte', type: 'String', value: 'NVIDIA Quadro K2200' }]);
    });

    describe('available', () => {
      it.each([
        [undefined, undefined, false],
        [false, undefined, false],
        [true, undefined, true],
        [undefined, false, false],
        [false, false, false],
        [true, false, false],
        [true, true, true],
      ])('should calculate availability=%s inStock=%s -> %s', (availability, inStock, expected) => {
        const data = { availability, inStock } as ProductData;
        expect(productMapper.fromData(data).available).toEqual(expected);
      });
    });
  });

  describe('fromStubData', () => {
    it('should throw an error when stub data has no sku attribute', () => {
      expect(() => productMapper.fromStubData({} as ProductDataStub)).toThrowErrorMatchingInlineSnapshot(
        `"cannot construct product stub without SKU"`
      );
    });

    it('should construct a stub when supplied with an API response', () => {
      const stub = productMapper.fromStubData({
        attributes: [{ name: 'sku', value: 'productSKU', type: 'String' }],
        title: 'productName',
        description: 'productDescription',
        attributeGroup: {
          name: 'attrGroup',
          attributes: [
            {
              name: 'attrName',
              type: 'Boolean',
              value: true,
            },
          ],
        },
      });
      expect(stub).toBeTruthy();
      expect(stub.name).toEqual('productName');
      expect(stub.shortDescription).toEqual('productDescription');
      expect(stub.sku).toEqual('productSKU');
      verify(imageMapper.fromImages(anything())).once();
    });

    it('should construct a stub when supplied with a complex API response', () => {
      const stub = productMapper.fromStubData({
        attributes: [
          { name: 'sku', value: '7912057' },
          { name: 'image', value: '/assets/product_img/a.jpg' },
          {
            name: 'listPrice',
            value: { currency: 'USD', type: 'Money', value: 132.24 },
          },
          {
            name: 'salePrice',
            value: { currency: 'USD', type: 'Money', value: 132.24 },
          },
          { name: 'availability', value: true },
          { name: 'manufacturer', value: 'Kodak' },
          { name: 'minOrderQuantity', value: { value: 5 } },
          { name: 'packingUnit', value: 'pcs.' },
          { name: 'inStock', value: false },
        ] as Attribute[],
        description: 'EasyShare M552, 14MP, 6.858 cm (2.7 ") LCD, 4x, 28mm, HD 720p, Black',
        title: 'Kodak M series EasyShare M552',
        attributeGroup: {
          name: 'attrGroup',
          attributes: [
            {
              name: 'attrName',
              type: 'Boolean',
              value: true,
            },
          ],
        },
      });

      expect(stub).toMatchSnapshot();
    });

    describe('available', () => {
      it.each([
        [undefined, undefined, false],
        [false, undefined, false],
        [true, undefined, true],
        [undefined, false, false],
        [false, false, false],
        [true, false, false],
        [true, true, true],
      ])('should calculate availability=%s inStock=%s -> %s', (availability, inStock, expected) => {
        const data = {
          attributes: [
            { name: 'sku', value: '123' },
            { name: 'availability', value: availability },
            { name: 'inStock', value: inStock },
          ],
        } as ProductDataStub;
        expect(productMapper.fromStubData(data).available).toEqual(expected);
      });
    });
  });

  describe('fromProductBundleData', () => {
    it('should map bundle stubs to sku and quantity', () => {
      expect(
        productMapper.fromProductBundleData([
          {
            attributes: [{ name: 'quantity', value: { value: 1 } }],
            uri: 'inSPIRED-inTRONICS-Site/-/products/201807191',
          },
          {
            attributes: [{ name: 'quantity', value: { value: 2 } }],
            uri: 'inSPIRED-inTRONICS-Site/-/products/201807192',
          },
          {
            attributes: [{ name: 'quantity', value: { value: 1 } }],
            uri: 'inSPIRED-inTRONICS-Site/-/products/201807193',
          },
        ] as Link[])
      ).toMatchInlineSnapshot(`
        Array [
          Object {
            "quantity": 1,
            "sku": "201807191",
          },
          Object {
            "quantity": 2,
            "sku": "201807192",
          },
          Object {
            "quantity": 1,
            "sku": "201807193",
          },
        ]
      `);
    });
  });

  describe('parseSKUfromURI()', () => {
    it.each([
      'site/products/123',
      'products/123',
      'site/products/123?test=dummy',
      'products/123?test=dummy',
      'site/products;spgid=dfds/123',
      'products;spgid=dfds/123',
      'site/products;spgid=dfds/123?test=dummy',
      'products;spgid=dfds/123?test=dummy',
      'site/products;spgid=dfds/123?test=dummy&test2=dummy',
      'products;spgid=dfds/123?test=dummy&test2=dummy',
    ])(`should parse correct sku when given '%s'`, uri => {
      expect(ProductMapper.parseSKUfromURI(uri)).toEqual('123');
    });
  });

  it('should find default variation for master product', () => {
    const variations = [
      { uri: 'inSPIRED-inTRONICS-Site/-/products/111' },
      {
        attributes: [{ name: 'defaultVariation', type: 'Boolean', value: true }],
        uri: 'inSPIRED-inTRONICS-Site/-/products/222',
      },
      { uri: 'inSPIRED-inTRONICS-Site/-/products/333' },
    ] as Link[];

    const result = ProductMapper.findDefaultVariation(variations);
    expect(result).toBe('222');
  });
});
