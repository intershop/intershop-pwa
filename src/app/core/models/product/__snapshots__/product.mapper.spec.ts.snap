// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Mapper fromStubData should construct a stub when supplied with a complex API response 1`] = `
Object {
  "attributeGroups": Object {
    "attrGroup": Object {
      "attributes": Array [
        Object {
          "name": "attrName",
          "type": "Boolean",
          "value": true,
        },
      ],
    },
  },
  "available": false,
  "completenessLevel": 2,
  "defaultCategoryId": undefined,
  "failed": false,
  "images": Array [
    Object {
      "effectiveUrl": "http://www.example.org/assets/product_img/a.jpg",
      "imageActualHeight": undefined,
      "imageActualWidth": undefined,
      "name": "front M",
      "primaryImage": true,
      "type": "Image",
      "typeID": "M",
      "viewID": "front",
    },
    Object {
      "effectiveUrl": "http://www.example.org/assets/product_img/a.jpg",
      "imageActualHeight": undefined,
      "imageActualWidth": undefined,
      "name": "front S",
      "primaryImage": true,
      "type": "Image",
      "typeID": "S",
      "viewID": "front",
    },
  ],
  "listPrice": Object {
    "currency": "USD",
    "type": "Money",
    "value": 132.24,
  },
  "longDescription": undefined,
  "manufacturer": "Kodak",
  "maxOrderQuantity": 100,
  "minOrderQuantity": 5,
  "name": "Kodak M series EasyShare M552",
  "packingUnit": "pcs.",
  "promotionIds": Array [],
  "readyForShipmentMax": undefined,
  "readyForShipmentMin": undefined,
  "roundedAverageRating": 0,
  "salePrice": Object {
    "currency": "USD",
    "type": "Money",
    "value": 132.24,
  },
  "shortDescription": "EasyShare M552, 14MP, 6.858 cm (2.7 \\") LCD, 4x, 28mm, HD 720p, Black",
  "sku": "7912057",
  "stepOrderQuantity": 1,
  "type": "Product",
}
`;
